### PR TITLE
Add tempered SMC

### PR DESCRIPTION
**Pull request by AdrienCorenflos**
This is not meant to work yet, but should have most of the features so that we can discuss.

Inspired by https://github.com/AdrienCorenflos/tempered-smc-optimization

### CODE REVIEW

**Comment by AdrienCorenflos** [comment]
Related issue: https://github.com/blackjax-devs/blackjax/issues/41

File: blackjax/tempered_smc.py
Comment (rlouf) [inline blackjax/tempered_smc.py:44]: Here I am torn between your solution of passing `mcmc_parameters` as an argument to this function, or the following:

```python
mcmc_kernel_factory = lambda potential: mcmc.new_kernel(potential_fn, parameters)

def kernel(potential_fn, mcmc_kernel_factory, new_mcmc_state, resampling_method, mcmc_step):
    raise NotImplementedError
```

File: blackjax/tempered_smc.py
Comment (rlouf) [inline blackjax/tempered_smc.py:94]: How about?

```python
tempered_potential_fn = lambda position: -state.lmbda * potential_fn(position)
kernel = mcmc_kernel_factory(tempered_potential_fn)
```

File: blackjax/tempered_smc.py
Comment (rlouf) [inline blackjax/tempered_smc.py:105]: Let me try to simplify this a little bit:

```python
mcmc_key, resampling_key = jax.random.split(rng_key, 2)

def mcmc_step(states, rng_key):
    keys = jax.random.split(rng_key, num_particules)
    states, info = jax.vmap(kernel)(states)
    return states, (states, info)

keys = jax.random.split(mcmc_key, num_mcmc_steps)
initial_states = jax.vmap(new_mcmc_state, in_axes=(0, None))(state.particles, tempered_potential_fn)
last_states, _ = jax.lax.scan(mcmc_step, initial_states, keys)
```

File: blackjax/tempered_smc.py
Comment (rlouf) [inline blackjax/tempered_smc.py:25]: We should propagate the info that comes from the MCMC algorithm, in a `mcmc_info` field. Can be useful for debugging.

File: blackjax/tempered_smc.py
Comment (rlouf) [inline blackjax/tempered_smc.py:34]: I don't see this used anywhere

File: blackjax/tempered_smc.py
Comment (rlouf) [inline blackjax/tempered_smc.py:41]: I would leave this at the end of the file

**Review by rlouf — CHANGES_REQUESTED** [review]
You got the idea ! I made some suggestions as to how the code might be simplified.

File: blackjax/tempered_smc.py
Comment (AdrienCorenflos) [reply blackjax/tempered_smc.py:70]: It's not exactly propagated is it? lambda is, but delta is used to increment lambda only.

File: blackjax/tempered_smc.py
Comment (AdrienCorenflos) [reply blackjax/tempered_smc.py:105]: I am not a fan of splitting keys within a loop, it is less efficient for very little added visibility. Also in the case of tempered SMC, there is no reason to store the intermediate states of the MCMC chain like you just did, (except for waste free I guess, but the implementation would be wildly different)

File: blackjax/tempered_smc.py
Comment (AdrienCorenflos) [reply blackjax/tempered_smc.py:25]: np

File: blackjax/tempered_smc.py
Comment (AdrienCorenflos) [reply blackjax/tempered_smc.py:34]: Yup, it's a copy paste residual

File: blackjax/tempered_smc.py
Comment (AdrienCorenflos) [reply blackjax/tempered_smc.py:41]: OK, I tend to prefer having auxiliary functions at the beginning, but your repo your rules!

File: blackjax/tempered_smc.py
Comment (AdrienCorenflos) [reply blackjax/tempered_smc.py:44]: I prefer your version. It makes the interface cleaner

File: blackjax/tempered_smc.py
Comment (AdrienCorenflos) [reply blackjax/tempered_smc.py:105]: More interestingly what do you think of the vmap here. Should it be the responsibility of the user (this way we allow for manual batching), or abstracted away in the code like I've done? It's pretty hard to catch the fact that the kernel may not be batched, so if we leave it as a user responsibility, returning readable errors saying "hey you forgot to batch" is probably gonna be impossible

File: blackjax/tempered_smc.py
Comment (rlouf) [reply blackjax/tempered_smc.py:41]: I don't like my novels to start with all the subplots, and I end up reading code a lot. But that's a matter of taste, I guess.

File: blackjax/tempered_smc.py
Comment (AdrienCorenflos) [reply blackjax/tempered_smc.py:70]: But I guess this decision can come at the end of the dev work as it would be a minimal change anyway, so maybe we should keep it for later when the rest of the design is clearer? Or is the reason why you want it in the state specific to the fact you want one_step to only take key + state?

Comment (AdrienCorenflos) [reply blackjax/tempered_smc.py:70]: the pb with having delta on the state is that it asks the user to manipulate the state in between two steps, essentially doing smthg like:
```python 
particles, weights, delta, lamda = state
state = State(particles, weights, new_delta, lamda)
```
which feels a bit wasteful for no reason

File: blackjax/tempered_smc.py
Comment (rlouf) [reply blackjax/tempered_smc.py:105]: - I haven't noticed much difference between splitting all at once and splitting within the loop, but I haven't tried in a while.
- We'll come to users batching themselves later I think; you wouldn't batch NUTS like you do HMC for instance, since NUTS chains can take very different number of steps at each round. For now we can batch inside as I don't know what the API will look like.

**Comment by AdrienCorenflos** [comment]
Thanks for the comments overall @rlouf I'm working on some paper code today but will continue the implementation soonish

File: blackjax/tempered_smc.py
Comment (rlouf) [reply blackjax/tempered_smc.py:44]: And this pattern of partially applying comes back often in the library, eg for adaptation.

File: blackjax/tempered_smc.py
Comment (AdrienCorenflos) [reply blackjax/tempered_smc.py:94]: That depends how much of a PEP purist you want to be. Assigning lambda functions to variables is now advised against as it messes up with the stack trace at debug time.

Comment (rlouf) [reply blackjax/tempered_smc.py:94]: I'm a purist within reason. This specific check is disabled on this repo. Since the codebase is purely functional, disallowing lambda function assignment would make the code harder to read.

File: blackjax/tempered_smc.py
Comment (rlouf) [reply blackjax/tempered_smc.py:70]: Actually, could you add the instruction where you change the value of  `delta`? I need the full picture to be able to judge.

**Comment by AdrienCorenflos** [comment]
@rlouf I added a function for choosing delta (non ideal, but will do). The resulting code is then something like 

```python
from ... import ...

target_ess = 0.75

smc_kernel = kernel(potential_fn, ...)

def tempering_body(carry):
    key, smc_state = carry
    key, subkey = split(key)
    delta = ess_solver(potential_fn, smc_state, target_ess)
    smc_state = smc_kernel(subkey, smc_state, next_delta)
    return key, smc_state

def tempering_stop_cond(carry):
    _, smc_state = carry
    return smc_state.lamda < 1

_, final_state = jax.lax.while_loop(tempering_stop_cond, tempering_body, ...)
```

**Comment by rlouf** [comment]
This is looking very good! I'm confident we'll soon converge to something we can merge :)

File: blackjax/tempered_smc.py
Comment (rlouf) [inline blackjax/tempered_smc.py:93]: We would get rid of a lot of code by letting `particles` be the MCMC state. Since all MCMC states have a `position` argument, we could safely pass `position.state` to every function that only needs the position.

So this line would go away.

File: blackjax/tempered_smc.py
Comment (rlouf) [inline blackjax/tempered_smc.py:99]:
```python
log_weights = -delta * jax.vmap(potential_fn, in_axes=[0])(particles.position)
```

File: blackjax/tempered_smc.py
Comment (rlouf) [inline blackjax/tempered_smc.py:97]: goes away

File: blackjax/tempered_smc.py
Comment (rlouf) [inline blackjax/tempered_smc.py:94]:
```python
particles, _ = jax.lax.scan(...)
```

(we'll think about what we pass as info later)

File: blackjax/tempered_smc.py
Comment (rlouf) [inline blackjax/tempered_smc.py:230]: Could go in `inference/smc/resampling.py` since there will likely be several resampling schemes in the future. I would move everything hmc related to `inference/hmc`.

File: blackjax/tempered_smc.py
Comment (rlouf) [inline blackjax/tempered_smc.py:79]: s/pytree/position?

File: blackjax/tempered_smc.py
Comment (rlouf) [inline blackjax/tempered_smc.py:118]: Everything ess-related could go in a `inference/smc/ess.py`

File: tests/test_tempered_smc.py
Comment (rlouf) [reply tests/test_tempered_smc.py:29]: Actually, how about computing the tempered potential in this loop as well so the SMC kernel above is very general? In which case you could move it to `inference/smc/kernel.py`.

The general SMC state would reduce to the MCMC state and here you could create a TemperedSMC state that includes the SMC state and delta.

File: blackjax/tempered_smc.py
Comment (AdrienCorenflos) [reply blackjax/tempered_smc.py:84]: okok

File: tests/test_tempered_smc.py
Comment (AdrienCorenflos) [reply tests/test_tempered_smc.py:29]: Yeah I was thinking about adding a level in the module hierarchy too to be able to split responsibilities.
Plan was indeed to have the delta calculation inside the tempering, but I put it outside as you wanted to see what it would be then. More than happy to move it back inside.

File: blackjax/tempered_smc.py
Comment (AdrienCorenflos) [reply blackjax/tempered_smc.py:93]: ok

File: blackjax/tempered_smc.py
Comment (AdrienCorenflos) [reply blackjax/tempered_smc.py:99]: ok

File: blackjax/tempered_smc.py
Comment (AdrienCorenflos) [reply blackjax/tempered_smc.py:97]: ok

File: blackjax/tempered_smc.py
Comment (AdrienCorenflos) [reply blackjax/tempered_smc.py:94]: agreed re the indo

File: blackjax/tempered_smc.py
Comment (AdrienCorenflos) [reply blackjax/tempered_smc.py:79]: ok

File: blackjax/tempered_smc.py
Comment (AdrienCorenflos) [reply blackjax/tempered_smc.py:118]: ok

**Comment by AdrienCorenflos** [comment]
I think we are in agreement. I was a bit unhappy having everything in the same file anyway

File: tests/test_tempered_smc.py
Comment (rlouf) [reply tests/test_tempered_smc.py:28]: The code structure would look something like

```python
def tempered_smc(...):
    # all the tempering logic here
    particles, info = general_smc_kernel(...)
    return particles, info
```

**Comment by AdrienCorenflos** [comment]
@rlouf  What do you think? I guess we are getting closer?

**Comment by rlouf** [comment]
We're getting really close.

**Comment by rlouf** [comment]
Let me know when the tests pass and I'll have a closer look. It would also be nice to showcase the algorithm in a notebook, show a situation where it really shines as opposed to HMC alone.

**Comment by AdrienCorenflos** [comment]
I forgot to take the prior into consideration in the code, hence the garbage numbers (it is arbitrary in the optimisation context, which is what my code was doing originally, so I didn't think about it). Need to think about it, but probably won't happen before Monday.

**Review by AdrienCorenflos — COMMENTED** [review]
> Let me know when the tests pass and I'll have a closer look. It would also be nice to showcase the algorithm in a notebook, show a situation where it really shines as opposed to HMC alone.

Example will probably be a good old multimodal distribution.

**Comment by rlouf** [comment]
No problem, take your time.

And yes for the multimodal distribution! I'm curious to see how this performs compared to parallel tempering with HMC.

**Comment by AdrienCorenflos** [comment]
Apart from cosmetics, this should do I believe!

**Comment by AdrienCorenflos** [comment]
@rlouf I don't trust the tests failures not gonna lie. Imports are well sorted.

**Comment by AdrienCorenflos** [comment]
> @rlouf I don't trust the tests failures not gonna lie. Imports are well sorted.

Looks like this is due to this isort thing. This doesn't feel like a standard python practice, what is it exactly/what's the purpose?

**Comment by rlouf** [comment]
This looks great @AdrienCorenflos ! I'll find some time to dive in your code this week.

The purpose is normalizing the code as much as possible to make it easier to read. It's useful on codebases on which more than one person works, and pretty standard afaik.

**Comment by rlouf** [comment]
PS: could you give me [push access](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to your branch? That way I can make cosmetic changes without bothering you. For everything else than cosmetic I'll leave comments and questions here.

**Comment by AdrienCorenflos** [comment]
> This looks great @AdrienCorenflos ! I'll find some time to dive in your code this week.
> 
> The purpose is normalizing the code as much as possible to make it easier to read. It's useful on codebases on which more than one person works, and pretty standard afaik.

Honestly first time I see it as a test on an open source package. Doesn't seem that well supported at least by pycharm

**Comment by AdrienCorenflos** [comment]
Err, all these tests pass locally...

**Comment by AdrienCorenflos** [comment]
> PS: could you give me [push access](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to your branch? That way I can make cosmetic changes without bothering you. For everything else than cosmetic I'll leave comments and questions here.

You already have it
image

File: blackjax/inference/smc/ess.py
Comment (AdrienCorenflos) [reply blackjax/inference/smc/ess.py]: This will be stable I believe, I removed the max element from the logw already. Also can't use logaddexp here with these. maybe 2 * logsumexp(logw) - logsumexp(2 logw), but it is wasteful, and less precise, as it does the same thing as I did but doesn't take into account the fact that max(logw) simplifies.

File: blackjax/inference/smc/resampling.py
Comment (AdrienCorenflos) [reply blackjax/inference/smc/resampling.py]: added residual this morning :)

File: blackjax/tempered_smc.py
Comment (AdrienCorenflos) [reply blackjax/tempered_smc.py:29]: Feel free to change the cosmetics the way you want.

Comment (AdrienCorenflos) [reply blackjax/tempered_smc.py:29]: I kept it like this to also keep the SMC independent so as to be able to implement sampling of other forms (SSM, etc), later on.

File: blackjax/inference/smc/smc.py
Comment (AdrienCorenflos) [reply blackjax/inference/smc/smc.py]: I don't see how this is complicated as it is using your high level interface directly.

**Comment by rlouf** [comment]
Heu don't worry about the imports for now, I'll fix it. You need to make sure to use the `--profile=black` flag with `isort`.

**Comment by AdrienCorenflos** [comment]
I've put in a somewhat illustrative example. You can make it fairer it by running parallel chains for HMC and NUTS, but I felt it was a bit unecessary.

**Comment by AdrienCorenflos** [comment]
@rlouf just checking in to see when we can wrap this up and call it a success :D

**Comment by rlouf** [comment]
Sorry, I've been traveling lately and been busy with other things. I will go over the PR next week so we don't lose too much momentum :)

**Comment by rlouf** [comment]
I have done a little refactoring (mostly to follow more closely the existing naming conventions/code structure) at the higher level and will make another pass next week. So far so good!

**Comment by rlouf** [comment]
I removed the `SMCState` NamedTuple as it only contained the particles. I changed `test_smc` and `test_tempered_smc` accordingly, but I might have done something wrong as they now all break. Here is what is left to do for me on this part:

- [ ] Fix the broken tests
- [ ] Add a `SamplerState` type and use instead of `PyTree` for particles
- [ ] Understand what's going on with the separate prior

Regarding the high level level algorithms, `fixed_schedule_tempered_smc` is not needed. Indeed, we could write the base kernel so it has the following signature:

```python
def one_step(rng_key, particles, lambda):
```

`fixed_schedule_tempered_smc` is then obtained by using `jax.lax.scan` to scan over the array of lambda values. `adaptive_tempered_smc` is then obtained by calling `one_step` with the computed value of `lambda` at every step.

**Comment by AdrienCorenflos** [comment]
> I removed the `SMCState` NamedTuple as it only contained the particles. I changed `test_smc` and `test_tempered_smc` accordingly, but I might have done something wrong as they now all break. Here is what is left to do for me on this part:
> 
> * [ ]  Fix the broken tests
> * [ ]  Add a `SamplerState` type and use instead of `PyTree` for particles
> * [ ]  Understand what's going on with the separate prior
> 
> Regarding the high level level algorithms, `fixed_schedule_tempered_smc` is not needed. Indeed, we could write the base kernel so it has the following signature:
> 
> ```python
> def one_step(rng_key, particles, lambda):
> ```
> 
> `fixed_schedule_tempered_smc` is then obtained by using `jax.lax.scan` to scan over the array of lambda values. `adaptive_tempered_smc` is then obtained by calling `one_step` with the computed value of `lambda` at every step.

To explain the separate prior, it comes from the fact that you only temper the likelihood term. A good way to see that there is no way around it is essentially to ask yourself what would happen for $\lambda=0$ otherwise.

**Comment by AdrienCorenflos** [comment]
Also why do you want to move the tempering schedule out of the algo? From my POV, it feels a bit like if you were asking the users to handle the symplectic integrator in the sampling loop too.

**Comment by rlouf** [comment]
Thanks for fixing everything I broke ! Looks like there are still a couple errors because I forgot to correct the paths in the files that define the adaptation algorithms.

1. Yes I re-read Chopin's book; I am so used to HMC and never separating the likelihood and prior that it kind of surprised me.
2. Yes we should make sure that the notebooks can run after every change. We should open a new issue for that.
3. I don't see how providing the array before iterating _simplifies_ anything. First you have to keep track of a number of iterations, which is awkward. You also need to keep track of how many times you've applied the kernel to make sure you don't go past the size of the array of temperatures.
All you are doing is "consuming" an array of temperatures, using `scan` does exactly that and you don't have to bother with indexing.

**Comment by AdrienCorenflos** [comment]
So what are you saying we do? Get the ess solver out and somehow ask the
user to configure it themselves to then pass it to the sampling loop?

**Comment by AdrienCorenflos** [comment]
IMO if you do that you just make the algo unusable for people who aren't
already experts of SMC and of annealing. Or maybe I don't have enough
perspective and you have a good idea of how to make this somehow user
friendly?

**Comment by rlouf** [comment]
From what I understood, finding a "good" temperature profile is the hardest thing about SMC. We can thus safely assume that people who are not familiar with annealing or SMC will use `adaptive_tempered_smc`. We thus leave this algorithm as is.

The important difference between `adaptive_tempered_smc` and the original version of `fixed_schedule_tempered_smc` is that in the latter the kernel holds a state which determines the maximum number of iterations. This is what bothers me the most, because it is neither idiomatic JAX code (using `scan` is recommended instead), nor is it justified. I think it is better to let users iterate over the array of temperatures themselves.

@junpenglao what do you think?

**Comment by AdrienCorenflos** [comment]
Ok I can get behind getting rid of the fixed schedule algo. I was worried
you wanted to get rid of the adaptive one.

**Comment by rlouf** [comment]
Great, I'll fix the remaining tests and formatting. Then we will wait until #66 is merged and add a new test for this algorithm before merging.

**Comment by rlouf** [comment]
I did some further refactoring. In particular, we don't need to pass `new_mcmc_state` function if we work with the RWMH/HMC/etc states directly. There might be some nitpicking stuff left. Now let's wait for #66 to be merged so we can add a new test to check sampling accuracy.

**Comment by rlouf** [comment]
I just rebased on the main branch. All we need now is a test like [this one](https://github.com/AdrienCorenflos/blackjax/blob/f71e8c38dcce86217144cf7c4434bdd658f0e2fe/tests/test_sampling.py#L212) to further check the validity of the samples. Then we're good to merge!

**Comment by rlouf** [comment]
You can test using high dimension gaussian, like for HMC and NUTS.

**Comment by AdrienCorenflos** [comment]
> You can test using high dimension gaussian, like for HMC and NUTS.

Done.

**Comment by rlouf** [comment]
Great job @AdrienCorenflos! I made some minor modifications (such as types) and made sure all tests pass. Thank you!

**Comment by AdrienCorenflos** [comment]
Yeeeah

**Pull request merged**
